utils/src/resource_location.cpp

```diff
@@ -356,81 +363,79 @@ resource_location resource_location::lexically_relative(const resource_location&
 
 bool resource_location::lexically_out_of_scope() const
 {
-    return m_uri == std::string_view("..") || m_uri.starts_with("../") || m_uri.starts_with("..\\");
+    std::string_view uri = get_uri();
+    return uri == std::string_view("..") || uri.starts_with("../") || uri.starts_with("..\\");
 }
 
-resource_location& resource_location::join(std::string_view other)
+resource_location& resource_location::join(std::string_view other) { return *this = join(std::move(*this), other); }
+
+resource_location resource_location::join(resource_location rl, std::string_view other)
 {
     if (utils::path::is_uri(other))
-        m_uri = other;
+        return resource_location(other);
     else if (other.starts_with("/"))
     {
-        auto dis_uri = utils::path::dissect_uri(m_uri);
+        auto dis_uri = utils::path::dissect_uri(rl.get_uri());
         dis_uri.path = other;
-        m_uri = utils::path::reconstruct_uri(dis_uri);
+        return resource_location(utils::path::reconstruct_uri(dis_uri));
     }
     else
-        uri_append(m_uri, other);
-
-    return *this;
+    {
+        std::string uri = rl.get_uri();
+        uri_append(uri, other);
+        return resource_location(std::move(uri));
+    }
 }
 
-resource_location resource_location::join(resource_location rl, std::string_view other)
+resource_location& resource_location::replace_filename(std::string_view other)
 {
-    rl.join(other);
-
-    return rl;
+    return *this = replace_filename(std::move(*this), other);
 }
 
-resource_location& resource_location::replace_filename(std::string_view other)
+resource_location resource_location::replace_filename(resource_location rl, std::string_view other)
 {
-    if (!utils::path::is_uri(m_uri))
+    if (!utils::path::is_uri(rl.get_uri()))
```

Comment:
A local variable `auto uri = rl.get_uri()` can probably be created instead of calling `rl.get_uri()` twice. 

Same can be considered for lines 549 - 557/

Reply:
That would force an extra copy, wouldn't it?

Reply:
Why or extra copy of what? 

`get_uri` returns `const std::string&`. What am I missing?

Reply:
if `is_uri(rl.get_uri())` the `auto uri = rl.get_uri()` would not be used by the rest of the code.

Reply:
There's also an instance of `rl.get_uri()` call outside of the if branch on line 406

Reply:
yes, but that's `const std::string&`->`std::string_view`